### PR TITLE
core: services: html: Update service to use fastapi and html files

### LINKED ISSUE #206: Helper: Add REST API
Add REST API to use it for frontend

### PR DESCRIPTION
**Pull request by patrickelectric**
image

Fix #206

### CODE REVIEW

File: core/services/helper/html/index.html [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+              function () {
+                this.requestData()
+              }.bind(this),
+              5000
+            )
+          },
+          requestData: function () {
+            console.debug('Requesting status..')
+            axios.get('/v1.0/web_services').then(
+              function (response) {
```
Comment (Williangalvani) [inline core/services/helper/html/index.html:94]:
```suggestion
              response => {
```
If you use an [arrow function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions) you can drop the `.bind(this)`

Comment (patrickelectric) [reply core/services/helper/html/index.html]: Done.

File: core/services/helper/html/index.html [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        methods: {
+          createUrlHyperlink: function (url) {
+            return `<a href="${url}">${url}</a>`
+          },
+          createWebpageUrl: function (port, path = '') {
+            return `${window.location.protocol}//${
+              window.location.host.split(':')[0]
+            }:${port}${path}`
+          },
+          createWebpageHyperlink: function (port, path = '') {
```
Comment (Williangalvani) [inline core/services/helper/html/index.html:80]:
```suggestion
          createWebpageHyperlink(port, path = '') {
```

"function" is unnecessary here

File: core/services/helper/html/index.html [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            setInterval(
+              () => {
+                this.requestData()
+              },
+              5000
+            )
+          },
+          requestData: function () {
+            console.debug('Requesting status..')
+            axios.get('/v1.0/web_services').then(
```
Comment (Williangalvani) [inline core/services/helper/html/index.html:93]:
```suggestion
            axios.get('/latest/web_services').then(
```

File: core/services/helper/setup.py [hunk 4]
```diff
@@ -1,22 +1,74 @@
...
+            try:
+                # Create path of file
+                download_path = pathlib.Path.joinpath(static_folder, path)
+                download_path.mkdir(exist_ok=True)
+                name = url.split("/")[-1]
+                # Create path with file name to download
+                download_file_path = pathlib.Path.joinpath(download_path, name)
+                urllib.request.urlretrieve(url, str(download_file_path))
+            except Exception as error:
+                print(f"Unable to download, error: {error}")
```
Comment (Williangalvani) [inline core/services/helper/setup.py:54]:
```suggestion
                print(f"Unable to download {url}, error: {error}")
```

**Review by Williangalvani — CHANGES_REQUESTED** [review]

**Review by rafaellehmkuhl — CHANGES_REQUESTED** [review]
Could you also add a .gitignore file including the static folders?

**Comment by patrickelectric** [comment]
> Could you also add a .gitignore file including the static folders?

Are you taking about this ?
https://github.com/bluerobotics/companion-docker/blob/39e0957e1168a3600572ea4c36e8a1da1c9d5377/core/services/helper/html/.gitignore

File: core/services/helper/setup.py (hunk 4 above)
Comment (patrickelectric) [reply core/services/helper/setup.py:54]: Check line 44.

Comment (Williangalvani) [reply core/services/helper/setup.py:54]: lol. my bad

**Comment by rafaellehmkuhl** [comment]
> > Could you also add a .gitignore file including the static folders?
> 
> Are you taking about this ?
> https://github.com/bluerobotics/companion-docker/blob/39e0957e1168a3600572ea4c36e8a1da1c9d5377/core/services/helper/html/.gitignore

Oh, ok, didn't see that it was inside the html folder.

Wouldn't it be better to have it inside the service folder? Otherwise if we need to add new ignores that are not in the html folder we will have to add a new .gitignore or move that one. It is also how we are doing currently in all other services.

File: core/services/helper/html/index.html [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+              this.requestData()
+            }, 5000)
+          },
+          requestData () {
+            console.debug('Requesting status..')
+            axios.get('/latest/web_services').then((response) => {
+              // Sort services by port number
+              this.services = response.data.sort(
+                (first, second) => first.port - second.port
+              )
```
Comment (rafaellehmkuhl) [inline core/services/helper/html/index.html:94]: Those spaces between the function name and the parenthesis are intentional?
I thought eslint would blame those.

File: core/services/helper/main.py [hunk 6]
```diff
@@ -1,42 +1,51 @@
...
+            indent=2,
+            separators=(", ", ": "),
+        ).encode(self.charset)
+
+
+class ServiceInfo(BaseModel):
     valid: bool
     title: str
-    ui: str
+    documentation: str
```
Comment (rafaellehmkuhl) [inline core/services/helper/main.py:41]: `documentation_url` maybe?

File: core/services/helper/main.py [hunk 7]
```diff
@@ -77,11 +86,32 @@ def scan_ports() -> List[ServiceInfo]:
         return list(valid_services)
 
 
-if __name__ == "__main__":
+FASTAPI = FastAPI(
```
Comment (rafaellehmkuhl) [inline core/services/helper/main.py:89]: Python convention all-uppercase just for constants. Also FastAPI is the name of the library itself, and could led to misunderstoods. Could we rename that one to something else? The convention in FastAPI is to just call it "app".

File: core/services/helper/main.py [hunk 8]
```diff
@@ -77,11 +86,32 @@ def scan_ports() -> List[ServiceInfo]:
...
+    response_model=List[ServiceInfo],
+    summary="Retrieve web services found.",
+)
+@version(1, 0)
+def web_services() -> Any:
+    """REST API endpoint to retrieve web services running."""
+    return Helper.scan_ports()
+
+
+APP = VersionedFastAPI(
```
Comment (rafaellehmkuhl) [inline core/services/helper/main.py:107]: I would say that shadowing it here as "app" does no harm, but "versioned_app" or something like that would also be good.

File: core/services/helper/main.py [hunk 9]
```diff
@@ -77,11 +86,32 @@ def scan_ports() -> List[ServiceInfo]:
...
-        return r"<br>".join([service.as_html() for service in services])
 
-    bottle.run(host="0.0.0.0", port=PORT)
+@FASTAPI.get(
+    "/web_services",
+    response_model=List[ServiceInfo],
+    summary="Retrieve web services found.",
+)
+@version(1, 0)
+def web_services() -> Any:
```
Comment (rafaellehmkuhl) [inline core/services/helper/main.py:102]: Missing response_model here.

**Review by rafaellehmkuhl — CHANGES_REQUESTED** [review]

File: core/services/helper/html/index.html (hunk 5 above)
Comment (patrickelectric) [reply core/services/helper/html/index.html:94]: > Those spaces between the function name and the parenthesis are intentional?
> I thought eslint would blame those.

I just run eslint, the tests are ok.

File: core/services/helper/main.py (hunk 7 above)
Comment (patrickelectric) [reply core/services/helper/main.py:89]: Done.

File: core/services/helper/main.py (hunk 8 above)
Comment (patrickelectric) [reply core/services/helper/main.py:107]: Done.

File: core/services/helper/main.py (hunk 6 above)
Comment (patrickelectric) [reply core/services/helper/main.py:41]: Done.

**Comment by patrickelectric** [comment]
> Wouldn't it be better to have it inside the service folder? Otherwise if we need to add new ignores that are not in the html folder we will have to add a new .gitignore or move that one. It is also how we are doing currently in all other services.

Done.

File: core/services/helper/main.py (hunk 9 above)
Comment (patrickelectric) [reply core/services/helper/main.py:102]: Check here: https://github.com/bluerobotics/companion-docker/pull/215/files/48e5b6bfd35fc696521ded5c848e1e38d140bcea#diff-6dc41ee6b98e636040e7ed21f5849b814cf34e7e52b547111b81ce7db196984eR98

Comment (rafaellehmkuhl) [reply core/services/helper/main.py:102]: Broken link

Comment (patrickelectric) [reply core/services/helper/main.py:102]: check lines above this one.

File: core/services/helper/html/index.html (hunk 2 above)
Comment (patrickelectric) [reply core/services/helper/html/index.html]: Check #223 to avoid such pain in the future.

File: core/services/helper/main.py (hunk 9 above)
Comment (rafaellehmkuhl) [reply core/services/helper/main.py:102]: Cool

**Review by rafaellehmkuhl — APPROVED** [review]

File: core/services/helper/main.py (hunk 9 above)
Comment (rafaellehmkuhl) [reply core/services/helper/main.py:102]: Cool. Didn't see

**Review by Williangalvani — APPROVED** [review]
Lgtm. This may need a rebase due to ~the incident~ #223 .

**Pull request merged**